### PR TITLE
[#719] Allow adding additional persistent volumes to brokers

### PR DESCRIPTION
**Pull request by gaohoward**

### CODE REVIEW

File: api/v1beta1/activemqartemis_types.go [hunk 1]
```diff
@@ -370,6 +370,34 @@ type DeploymentPlanType struct {
 	// Specifies the Revision History Limit of the statefulset
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Revision History Limit"
 	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty"`
+	// Specifies External PersistentVolumeClaims for the broker pod
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="External PVCs"
+	ExternalPVCs []ExternalPVCType `json:"externalPvcs,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:375]: The array of `ExternalPVCType` could be replaced by a map witht the name as key and the `PersistentVolumeClaimSpec` as value to avoid the new struct type `ExternalPVCType`. Eventually this map could be added to the `ActiveMQArtemisSpec` struct with the name `persistentVolumeClaims`, i.e.
```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata: 
  name: artemis-role-map
spec:
  persistentVolumeClaims (map[string]corev1.PersistentVolumeClaimSpec):
    myclaim:
      accessModes:
        - ReadWriteOnce
      volumeMode: Filesystem
      resources:
        requests:
          storage: 8Gi
        ...
```

File: api/v1beta1/activemqartemis_types.go [hunk 2]
```diff
@@ -370,6 +370,34 @@ type DeploymentPlanType struct {
...
+	AccessModes []corev1.PersistentVolumeAccessMode `json:"accessModes,omitempty"`
+	// The Storage
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Storage"
+	Storage StorageType `json:"storage,omitempty"`
+	// The mount path
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Mount Path"
+	MountPath *string `json:"mountPath,omitempty"`
+}
+
+type CustomVolumeType struct {
```
[the remark this answers is not in the document]
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:393]: yeah, makes sense. will update. Thanks!

File: api/v1beta1/activemqartemis_types.go (hunk 1 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:375]: ok, will update. Thanks!

Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: I have add an option outside the persistentVolumeClaim to allow user to specify mountPath.

Comment (brusdev) [reply api/v1beta1/activemqartemis_types.go]: > I have add an option outside the persistentVolumeClaim to allow user to specify mountPath.

Users could use extraVolumeMounts to define mountPath and other mount options for a persistent volume claim, i.e.
```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata: 
  name: artemis-role-map
spec:
  persistentVolumeClaims (map[string]corev1.PersistentVolumeClaimSpec):
    myclaim:
      accessModes:
        - ReadWriteOnce
      volumeMode: Filesystem
      resources:
        requests:
          storage: 8Gi
    extraVolumes ([]corev1.Volume):
      - name: mypd
        persistentVolumeClaim:
          claimName: myclaim
    extraVolumeMounts ([]corev1.VolumeMount):
      - mountPath: "/var/www/html"
        name: mypd
        ...
```

Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: That shouldn't work. myclaim is a template, it will be translated to real name for each pod, .i.e myclaim-<cr name>-ss-0 for pod0 and so on. extraVolumes and extraVolumeMounts should referencing static volume names, and they are going to mount to all pods. (not one for each pod)

Comment (brusdev) [reply api/v1beta1/activemqartemis_types.go]: I see that we are talking of 2 different options:
1. one persistent volume claim shared with all broker pods;
2. one persistent volume claim template used to define a persistent volume claim for every broker pods;

Option 1 should work without issues. Do we need this option?
```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata: 
  name: artemis-role-map
spec:
  persistentVolumeClaims (map[string]corev1.PersistentVolumeClaimSpec):
    myclaim:
      accessModes:
        - ReadWriteOnce
      volumeMode: Filesystem
      resources:
        requests:
          storage: 8Gi
  deploymentPlan:
    extraVolumes ([]corev1.Volume):
      - name: mypd
        persistentVolumeClaim:
          claimName: myclaim
    extraVolumeMounts ([]corev1.VolumeMount):
      - mountPath: "/var/www/html"
        name: mypd
        ...
```

Option 2 could work in the same way as it works in [StatefulSetSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulsetspec-v1-apps). The operator could create a persistent volume claim for every broker pods using the template and link an extra volume mount to its related persistent volume claim template using the name, i.e. adding a new filed `volumeClaimTemplates` (a map with the name as key and the PersistentVolumeClaimSpec as value):
```
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata: 
  name: artemis-role-map
spec:
  volumeClaimTemplates (map[string]corev1.PersistentVolumeClaimSpec):
    myclaim:
      accessModes:
        - ReadWriteOnce
      volumeMode: Filesystem
      resources:
        requests:
          storage: 8Gi
  deploymentPlan:
    extraVolumeMounts ([]corev1.VolumeMount):
      - mountPath: "/var/www/html"
        name: myclaim
        ...
```

Reusing `extraVolumeMounts` for `volumeClaimTemplates` too has 3 benefits:
1.  Avoid new types;
2. Reuse an API that users already know, see [StatefulSetSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulsetspec-v1-apps);
3. Expose all [VolumeMount](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#volumemount-v1-core) fields;

File: docs/help/operator.md [hunk 3]
```diff
@@ -1029,3 +1029,104 @@ stringData:
...
+          claimName: my-pvc-0
+    extraVolumeMounts:
+      - name: mydata
+        mountPath: /opt/mydata
+```
+When deploying the above CR, the PVC volume will be mounted to path **/opt/mydata** in the broker container of both broker pods. The **extraVolumeMounts** is optional. If not specified a default mountPath is given based on the type of the volume, as shown in the following list:
+
+|type of volume|default mountPath|
+|---|---|
+|HostPath|/amq/extra/hostpaths/<volume.name>|
```
Comment (gtully) [inline docs/help/operator.md:1073]: this must be a pattern that we can describe, otherwise this list will be out of date very quickly. Just describing the pattern should be sufficient.

File: api/v1beta1/activemqartemis_types.go (hunk 1 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: I think both options deals with different use cases. We should give user those options. Regarding the templates I think you are right that use VolumeMount type to define the mount options for the template. I'll add that. thanks!

File: docs/help/operator.md (hunk 3 above)
Comment (gaohoward) [reply docs/help/operator.md:1073]: It makes sense. will update. Thanks!

**Comment by gaohoward** [comment]
@brusdev @gtully  branch updated. :)

File: docs/help/operator.md [hunk 4]
```diff
@@ -1050,3 +1050,78 @@ stringData:
...
+
+### Attaching extra volumes shared by all broker pods
+
+When you have some existing volumes to be used by brokers, you can use **spec.deploymentPlan.extraVolumes** and **spec.deploymentPlan.extraVolumeMounts** to configure a broker custom resource to mount those volumes. For example if you have a PersistentVolumeClaim type volume called **my-pvc-0** and it has 1G capacity with some useful data in it:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: my-pvc-0
```
Comment (brusdev) [inline docs/help/operator.md:1064]: the suffix `-0` could confuse
```suggestion
  name: my-pvc
```

File: controllers/activemqartemis_reconciler.go [hunk 5]
```diff
@@ -1554,14 +1566,43 @@ func addNewVolumeMounts(existingNames map[string]string, existing *[]corev1.Volu
...
 		volumeMounts = append(volumeMounts, persistentCRVlMnt...)
 	}
 
+	for _, volume := range customResource.Spec.DeploymentPlan.ExtraVolumes {
+		var volumeMount *corev1.VolumeMount = nil
+		for _, vm := range customResource.Spec.DeploymentPlan.ExtraVolumeMounts {
+			if vm.Name == volume.Name {
+				volumeMount = vm
+				if volumeMount.MountPath == "" {
+					volumeMount.MountPath = volumes.GetDefaultMountPath(&volume)
```
Comment (brusdev) [inline controllers/activemqartemis_reconciler.go:1583]: the name of a volume must be a DNS_LABEL and unique within the pod, see [Volume v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#volume-v1-core) so the base dire of the default mount path for extra volumes could be the same for all volume types, i.e. for an extra volume with the name `mydata`, the default path could be:
```
/amq/extra/volumes/mydata
```

In this way the implementation will not require updates to handle new volume types.

File: api/v1beta1/activemqartemis_types.go [hunk 6]
```diff
@@ -373,6 +373,17 @@ type DeploymentPlanType struct {
...
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volumes"
+	ExtraVolumes []corev1.Volume `json:"extraVolumes,omitempty"`
+	// Specifies mount options for extraVolumes
+	// +optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Mount Options"
+	ExtraVolumeMounts []*corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
+	// Specifies External PersistentVolumeClaims for the broker pod
+	// the key is mount path
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="External PVCs"
+	PersistentVolumeClaims map[string]corev1.PersistentVolumeClaimSpec `json:"persistentVolumeClaims,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:386]: Maybe reusing the same field name of the [StatefulSetSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulsetspec-v1-apps) struct could help users
```suggestion
	// Specifies Volume Claims Templates for the broker pods
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Volume Claims Templates"
	VolumeClaimTemplates map[string]corev1.PersistentVolumeClaimSpec `json:"volumeClaimTemplates,omitempty"`
```

File: controllers/activemqartemis_controller2_test.go [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+		AfterEachSpec()
+	})
+
+	Context("persistent volumes tests", Label("controller-2-test"), func() {
+		It("external volumes attach", func() {
+			if os.Getenv("USE_EXISTING_CLUSTER") == "true" {
+
+				By("create extra volume")
+				pvc, createdPvc := DeployCustomPVC("my-pvc-0", defaultNamespace, func(candidate *corev1.PersistentVolumeClaim) {
```
Comment (brusdev) [inline controllers/activemqartemis_controller2_test.go:96]: the suffix `-0` could cause confusion
```suggestion
				pvc, createdPvc := DeployCustomPVC("my-pvc", defaultNamespace, func(candidate *corev1.PersistentVolumeClaim) {
```

File: controllers/activemqartemis_controller2_test.go [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+					candidate.Spec.AccessModes = []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					}
+					candidate.Spec.Resources = corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse("2Mi"),
+						},
+					}
+				})
```
Comment (brusdev) [inline controllers/activemqartemis_controller2_test.go:213]: is this PVC used?

**Comment by brusdev** [comment]
#719 could be incorporate by #758 but I would not do that because the users would need a complex patch and internal knowledge for a use case that could quite common.

**Comment by gaohoward** [comment]
> #719 could be incorporate by #758 but I would not do that because the users would need a complex patch and internal knowledge for a use case that could quite common.

+1. I'll update the PR to adapt your last reviews. Thanks Dom!

File: docs/help/operator.md (hunk 4 above)
Comment (gaohoward) [reply docs/help/operator.md:1094]: +1

File: controllers/activemqartemis_reconciler.go (hunk 5 above)
Comment (gaohoward) [reply controllers/activemqartemis_reconciler.go:1639]: I was thinking to give the default mount path some information of the volume types that may be useful for users. However that's not necessary as user have the option to specify the path of their own choices. Having said that I do agree that use a common base path is much simpler! will update. Thanks.

File: api/v1beta1/activemqartemis_types.go (hunk 6 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:387]: +1

File: controllers/activemqartemis_controller2_test.go (hunk 7 above)
Comment (gaohoward) [reply controllers/activemqartemis_controller2_test.go:96]: +1

File: controllers/activemqartemis_controller2_test.go (hunk 8 above)
Comment (gaohoward) [reply controllers/activemqartemis_controller2_test.go:213]: yes it's used in test. "my-data" maps to the base name of the pvc template and is mounted for test to write some bytes and verify.

**Comment by gaohoward** [comment]
Hi @brusdev, just updated the PR. Pls review. Thanks!

File: api/v1beta1/activemqartemis_types.go [hunk 9]
```diff
@@ -377,6 +377,16 @@ type DeploymentPlanType struct {
...
+	// Specifies additional Volumes to be attached to the broker pods
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volumes"
+	ExtraVolumes []corev1.Volume `json:"extraVolumes,omitempty"`
+	// Specifies mount options for extraVolumes
+	// +optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Mount Options"
+	ExtraVolumeMounts []*corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
+	// Specifies Extra Volume Claims Templates for the broker pods
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volume Claims Templates"
+	ExtraVolumeClaimTemplates map[string]corev1.PersistentVolumeClaimSpec `json:"extraVolumeClaimTemplates,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:389]: The field `ExtraVolumeClaimTemplates` could be an array of [PersistentVolumeClaim](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#persistentvolumeclaim-v1-core) as in [statefulset](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulset-v1-apps). In this way the users could easily add the labels and annotations.
```suggestion
	ExtraVolumeClaimTemplates []corev1.PersistentVolumeClaim `json:"extraVolumeClaimTemplates,omitempty"`
```

File: api/v1beta1/activemqartemis_types.go [hunk 10]
```diff
@@ -377,6 +377,16 @@ type DeploymentPlanType struct {
 	// Specifies the container security context
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Container Security Context"
 	ContainerSecurityContext *corev1.SecurityContext `json:"containerSecurityContext,omitempty"`
+	// Specifies additional Volumes to be attached to the broker pods
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volumes"
+	ExtraVolumes []corev1.Volume `json:"extraVolumes,omitempty"`
+	// Specifies mount options for extraVolumes
+	// +optional
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Mount Options"
+	ExtraVolumeMounts []*corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:386]: The field `ExtraVolumeMounts` could be an array of [VolumeMount](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#volumemount-v1-core) as in the [container](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#container-v1-core).
```suggestion
	ExtraVolumeMounts []corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
```

**Review by brusdev — COMMENTED** [review]
@howardgao it LGTM, I have just proposed 2 minor refinements

File: api/v1beta1/activemqartemis_types.go (hunk 9 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:389]: +1

File: api/v1beta1/activemqartemis_types.go (hunk 10 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:386]: +1

File: api/v1beta1/activemqartemis_types.go (hunk 9 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:389]: @brusdev Hi Dom, I'm having issue after I made this field as list of PersistentVolumeClaim. I cannot make it right, when i deploy such a cr 
`
apiVersion: broker.amq.io/v1beta1
apiVersion: broker.amq.io/v1beta1
kind: ActiveMQArtemis
metadata:
  name: ex-aao
spec:
  deploymentPlan:
    size: 1
    extraVolumeClaimTemplates:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        name: mydata
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 10Mi
` 
I got the following error 
`Error from server (BadRequest): error when creating "test-broker.yaml": ActiveMQArtemis in version "v1beta1" cannot be handled as a ActiveMQArtemis: strict decoding error: unknown field "spec.deploymentPlan.extraVolumeClaimTemplates[0].metadata.name"
`
Wondering what's the correct way to configure the pvc's name.

Comment (brusdev) [reply api/v1beta1/activemqartemis_types.go:389]: @gaohoward I think this is an issue of the kustomize tool. We could create the `VolumeClaimTemplate` struct to emulate the array of [PersistentVolumeClaim](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#persistentvolumeclaim-v1-core) as in [statefulset](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#statefulset-v1-apps), i.e.
```
...
	// Specifies Extra Volume Claims Templates for the broker pods
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volume Claims Templates"
	ExtraVolumeClaimTemplates []VolumeClaimTemplate `json:"extraVolumeClaimTemplates,omitempty"`
}

type VolumeClaimTemplate struct {
	// Standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Metadata"
	ObjectMeta `json:"metadata,omitempty"`

	// spec defines the desired characteristics of a volume requested by a pod author.
	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#persistentvolumeclaims
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Spec"
	Spec corev1.PersistentVolumeClaimSpec `json:"spec,omitempty"`
}

type ObjectMeta struct {
	// Name must be unique within a namespace. Is required when creating resources, although
	// some resources may allow a client to request the generation of an appropriate name
	// automatically. Name is primarily intended for creation idempotence and configuration
	// definition.
	// Cannot be updated.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Name"
	Name string `json:"name,omitempty"`

	// Namespace defines the space within which each name must be unique. An empty namespace is
	// equivalent to the "default" namespace, but "default" is the canonical representation.
	// Not all objects are required to be scoped to a namespace - the value of this field for
	// those objects will be empty.
	//
	// Must be a DNS_LABEL.
	// Cannot be updated.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Namespace"
	Namespace string `json:"namespace,omitempty"`

	// Annotations is an unstructured key value map stored with a resource that may be
	// set by external tools to store and retrieve arbitrary metadata. They are not
	// queryable and should be preserved when modifying objects.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Annotations"
	Annotations map[string]string `json:"annotations,omitempty"`

	// Map of string keys and values that can be used to organize and categorize
	// (scope and select) objects. May match selectors of replication controllers
	// and services.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Labels"
	Labels map[string]string `json:"labels,omitempty"`
}
```

Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: +1

File: controllers/activemqartemis_reconciler.go [hunk 11]
```diff
@@ -2496,26 +2546,36 @@ func (reconciler *ActiveMQArtemisReconcilerImpl) NewStatefulSetForCR(customResou
...
+			pvcArray = append(pvcArray, *pvc)
+		}
 	}
 
-	for i := 0; i < arrayLength; i++ {
-		pvc = persistentvolumeclaims.NewPersistentVolumeClaimWithCapacityAndStorageClassName(namespacedName, capacity, namer.LabelBuilder.Labels(), storageClassName)
-		reconciler.applyTemplates(pvc)
+	for _, epvc := range customResource.Spec.DeploymentPlan.ExtraVolumeClaimTemplates {
+		nn := types.NamespacedName{Name: epvc.Name, Namespace: customResource.Namespace}
+		pvc = persistentvolumeclaims.NewPersistentVolumeClaim(nn, &epvc.Spec)
```
Comment (brusdev) [inline controllers/activemqartemis_reconciler.go:2578]: The function `persistentvolumeclaims.NewPersistentVolumeClaim` ignores the annotations and labels in epvc, the new parameters could be the namespace and the epvc, i.e.
```suggestion
		pvc = persistentvolumeclaims.NewPersistentVolumeClaim(customResource.Namespace, epvc)
```

File: api/v1beta1/activemqartemis_types.go [hunk 12]
```diff
@@ -377,6 +378,63 @@ type DeploymentPlanType struct {
...
+	// Namespace defines the space within which each name must be unique. An empty namespace is
+	// equivalent to the "default" namespace, but "default" is the canonical representation.
+	// Not all objects are required to be scoped to a namespace - the value of this field for
+	// those objects will be empty.
+	//
+	// Must be a DNS_LABEL.
+	// Cannot be updated.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Namespace"
+	Namespace string `json:"namespace,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:423]: The pods and the PersistentVolumeClaim have to be in the same namespace, the namespace field could cause confusion, it should be removed, i.e.
```suggestion
```

Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:423]: +1

File: controllers/activemqartemis_reconciler.go (hunk 11 above)
Comment (gaohoward) [reply controllers/activemqartemis_reconciler.go:2578]: +1

**Comment by gaohoward** [comment]
@brusdev Branch updated. Thanks!

**Review by brusdev — APPROVED** [review]
I executed the full test suite without test failures by using the fixes from #816 and #817.

**Comment by gaohoward** [comment]
> I executed the full test suite without test failures by using the fixes from #816 and #817.

@brusdev  Nice job!.  will update the PR. Thanks!

File: controllers/activemqartemis_controller2_test.go [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+
+	Context("persistent volumes tests", Label("controller-2-test"), func() {
+		It("external volumes attach", func() {
+			if os.Getenv("USE_EXISTING_CLUSTER") == "true" {
+
+				By("create extra volume")
+				pvc, createdPvc := DeployCustomPVC("my-pvc", defaultNamespace, func(candidate *corev1.PersistentVolumeClaim) {
+
+					candidate.Spec.AccessModes = []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteMany,
```
Comment (brusdev) [inline controllers/activemqartemis_controller2_test.go:99]: Running the full test-suite on an AWS OpenShift cluster causes the failure of the test `external volumes attach` because the default storage class doesn't support the access mode `ReadWriteMany`. That test failure could be easily fixed using the access mode `ReadWriteOnce`, i.e.
```suggestion
						corev1.ReadWriteOnce,
```

Comment (gaohoward) [reply controllers/activemqartemis_controller2_test.go:99]: ok will update. Thanks!

File: api/v1beta1/activemqartemis_types.go [hunk 14]
```diff
@@ -69,6 +69,7 @@ type ActiveMQArtemisSpec struct {
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Ingress Domain",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
 	IngressDomain string `json:"ingressDomain,omitempty"`
 	// Specifies the template for various resources that the operator controls
+	//+kubebuilder:pruning:PreserveUnknownFields
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:72]: The new marker `//+kubebuilder:pruning:PreserveUnknownFields` for `ResourceTemplates` doesn't seem related to this PR.
```suggestion
```

File: api/v1beta1/activemqartemis_types.go [hunk 15]
```diff
@@ -377,6 +378,52 @@ type DeploymentPlanType struct {
...
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Mount Options"
+	ExtraVolumeMounts []corev1.VolumeMount `json:"extraVolumeMounts,omitempty"`
+	// Specifies Extra Volume Claims Templates for the broker pods
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Extra Volume Claims Templates"
+	ExtraVolumeClaimTemplates []VolumeClaimTemplate `json:"extraVolumeClaimTemplates,omitempty"`
+}
+type VolumeClaimTemplate struct {
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Metadata"
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:395]: ObjectMeta is not a standard object's metadata
```suggestion
	// Specifies the desired metadata of a volume claim
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Metadata"
```

File: api/v1beta1/activemqartemis_types.go [hunk 16]
```diff
@@ -377,6 +378,52 @@ type DeploymentPlanType struct {
...
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// spec defines the desired characteristics of a volume requested by a pod author.
+	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#persistentvolumeclaims
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Spec"
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:400]: The description of the field Spec is not aligned with other field description
```suggestion
	// Specifies the desired characteristics of a volume claim
	// More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#persistentvolumeclaims
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Spec"
```

File: api/v1beta1/activemqartemis_types.go [hunk 17]
```diff
@@ -377,6 +378,52 @@ type DeploymentPlanType struct {
...
+
+type ObjectMeta struct {
+	// Name must be unique within a namespace. Is required when creating resources, although
+	// some resources may allow a client to request the generation of an appropriate name
+	// automatically. Name is primarily intended for creation idempotence and configuration
+	// definition.
+	// Cannot be updated.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Name"
+	Name string `json:"name,omitempty"`
```
Comment (brusdev) [inline api/v1beta1/activemqartemis_types.go:412]: The field name doesn't define any xDescriptors
```suggestion
	// Name must be unique within a namespace. Is required when creating resources, although
	// some resources may allow a client to request the generation of an appropriate name
	// automatically. Name is primarily intended for creation idempotence and configuration
	// definition.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names
	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Name",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
	Name string `json:"name,omitempty"`
```

File: api/v1beta1/activemqartemis_types.go (hunk 14 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go:72]: sorry I added it when debugging but forgot to delete it. will update.

File: api/v1beta1/activemqartemis_types.go (hunk 15 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: +1

File: api/v1beta1/activemqartemis_types.go (hunk 16 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: +1

File: api/v1beta1/activemqartemis_types.go (hunk 17 above)
Comment (gaohoward) [reply api/v1beta1/activemqartemis_types.go]: +1

**Review by brusdev — APPROVED** [review]
LGTM, I'll merge this PR after the executions of the PR checks.
